### ISSUE TITLE
Cross-encoder relevance arbiter (in-process travsr-rerank): wire into get_context + travsr ask — RFC-021

### ISSUE THREAD

**Issue by Abhishek5517** [Travsr-com/travsr#459]
> **Updated 2026-07-17 after Principal Architect end-to-end review (RFC-021 §15).** Delivery flipped from sidecar to **in-process**. The reranker now ships as a new pure-Rust crate `travsr-rerank` linked into the main binary — it does **not** depend on the travsr-embed sidecar (#8), so this fix is unblocked from the #6 ORT epic. See §15 for the full ruling/gap register.

## Summary

Add a **cross-encoder reranker** as the single **absolute, abstention-capable** relevance signal for `get_context` and `travsr ask`, replacing the token-geometry confidence gate that produces confident-salad. Implements **RFC-021 — Cross-Encoder Relevance Arbiter** (`docs/rfcs/RFC-021-cross-encoder-relevance-arbiter.md`), v1 = **in-process** (`travsr-rerank` crate, pure-Rust `tract`, CPU, deterministic).

## Why

Reproduced on this repo with a full arctic-embed-m index (embeddings ON):

```
query: "delete all user accounts and drop the database"
→ [retrieval: exact+lexical+semantic | coverage 3/5 | confidence: STRONG ]
  fn:Guard.drop (Rust Drop impl)      [score 1.00]
  fn:SqliteStore.delete_file           [score 0.82]
  ... 80 nodes
```

Root cause (RFC-021 §2): confidence is derived from **geometric** signals (BM25 term overlap + separately-encoded bi-encoder cosine), both token-overlap-driven — they cannot tell "the user's words are also symbol names" from "the user's intent matches this code." A **cross-encoder** encodes the pair *jointly* and yields an absolute `[0,1]` score with a real abstain floor.

## Architecture (ratified — RFC-021 §15)

- **New leaf crate `travsr-rerank`** — a `Reranker` trait + `TractReranker` (pure-Rust `tract`, CPU, **in-process**). Deterministic (fixed accumulation order → preserves "same query, same result"). Deps: `travsr-core` (`NodeId`) + `tract` only. **`travsr-core`/`travsr-retrieval` stay ML-free** (G3).
- **`travsr-mcp` depends on `travsr-rerank`.** A future sidecar-backed `Reranker` impl (travsr-embed #8) is a GPU/cloud variant behind the same trait — **not a dependency of this issue**.
- **Model:** `cross-encoder/ms-marco-MiniLM-L6-v2`, standard BERT, `model_fp16.onnx` = 45.6 MB. Ships **alongside the platform binary in GitHub Releases**, pulled by the existing binary-fetch (npm wrapper / brew) — present offline, default-on. `TRAVSR_NO_RERANK=1` → lexical fallback (G4).

## Two defects being closed

- **A — false STRONG via coincidental generic-verb anchors.** `seed.rs:595` `has_specific_anchor && coverage >= cov_strong => Strong`; `embed_agrees` (`seed.rs:520`) fooled by lexical-overlap-inflated cosine; the RFC-019 `confirm_anchor_floor` guard is only on the low-coverage branch (`seed.rs:604`).
- **B — per-node score is relative, not absolute.** `tools.rs:3149-3162` displays `raw / max_raw_score` → top node always `1.00`. `travsr ask` has its own instance: PPR × degree-damping in `AskRow.score` (`query.rs:299-306`).

### 1. `travsr-rerank` crate (new)
- `Reranker` trait: `rerank(query, &[candidate_text]) -> Vec<f32>` in `[0,1]`.
- `TractReranker`: load MiniLM-L6 fp16, WordPiece tokenizer, pair input `[CLS] q [SEP] cand [SEP]` + `token_type_ids`, 1-logit → sigmoid, batched (K ≤ 40, seq ≤ 256).
- **Runtime safety (G5):** dedicated bounded **blocking (HEAVY) pool**, never the async/MCP thread; `catch_unwind` around inference (panic → degrade that query to lexical gate, daemon survives); **fail-open** on model-load failure.
- **Cold start (G6):** background-warm at daemon start on a non-blocking thread (reuse the `embed_ready` lazy-init pattern that fixed the MCP-init hang); `rerank_budget_ms` circuit-breaker for contention.

### 2. Rerank hook in `build_seed_set` (seed.rs:896, after RRF at seed.rs:1096)
- Take top-K fused candidates (`rerank_topk`, default 30); candidate text = **signature-first** + skeleton, truncated to 256 tok (truncation drops body, not signature; reuse the `include_snippets` extractor) (G8).
- Rerank runs **after** `filter.allow` (RBAC-filtered seeds only — already true) (G8).
- Store `Seed.rerank_score: Option<f32>`; reorder.

### 3. Collapse `classify_confidence` (seed.rs:477) — NL queries only
```rust
match max_rerank_score {
    Some(r) if r >= STRONG_FLOOR => Confidence::Strong,
    Some(r) if r >= WEAK_FLOOR   => Confidence::Weak,
    Some(_)                       => Confidence::None,   // abstain — honest no-match
    None => classify_confidence_lexical_fallback(...),   // reranker off/warming/timeout
}
```
- **BLOCKER fix G1 — reranker arbitrates NL/conceptual queries ONLY.** When `build_seed_set` resolved a `SeedSource::Exact` anchor (the user named a real symbol), confidence is driven by the **deterministic exact match**; the reranker is **bypassed for the confidence decision** (may still reorder the tail). Prevents false-abstention on exact-symbol lookups (`GetWarningsForPod` etc.).
- **G2 — embeddings never touch confidence.** Embedding KNN is a candidate **source only** (raw cosine × kind_boost into RRF) → recall only. **Delete from the confidence path:** `Calibration{lo,hi}`, `semantic_veto`, `semantic_strong`, `embed_confirmed_specific`, `confirm_anchor_floor`, `disjoint_rescue_cos`, `anchor_neighborhood` contamination gate, and the embed-on/off branching. Confidence identical embed-on/off → kills the "some repos" inconsistency.
- **Floors (G7):** `STRONG_FLOOR`/`WEAK_FLOOR` measured from `bench/` (see §5); no provisional constants ship. Model version + floors pinned together in the model manifest (atomic bump — no calibration drift).

### 4. Per-node score = absolute rerank score (Defect B)
- `get_context` (`tools.rs:3149-3162`): display the absolute rerank score, not `raw/max_raw`. No forced `1.00`.
- `travsr ask` (`query.rs`): `AskRow.score` = the node's `rerank_score` for seed rows; structurally-expanded rows (`enrich_seeds_with_callers`/`dedup_adjacent_seeds`, `query.rs:253-256`) keep a PPR score **capped at the parent seed's rerank score** + tagged `[via: caller of …]`.

### 5. `travsr ask` parity (query.rs:191)
- `ask_query` already abstains on `Confidence::None` (`query.rs:240`) → becomes correct automatically. `ask "delete all user accounts and drop the database"` → `matched: false`.
- Add `AskPayload.confidence: &'static str` so the CLI prints `confidence: strong|weak` (VS Code graph panel hits this path via `:`-prefixed queries, `query.rs:197`).
- Cross-surface parity: `ask` and `get_context` return the **same** confidence label + abstain decision for the same query.

### 6. `bench/` negative set (G7)
- Add a labelled **negative set** (off-topic + generic-vocabulary queries incl. the reproduction case). Positives = on-topic symbol + conceptual queries. Choose floors at the ROC point maximising negative-abstention s.t. positive-recall ≥ target; record values + model version in the manifest.

### 7. `travsr-cli` — default-on bundling (G4)
- Bundle/fetch the fp16 model with the platform binary; no opt-in step; `TRAVSR_NO_RERANK=1` escape hatch.

## Acceptance criteria

- [ ] `travsr-rerank` crate with `Reranker` trait + `TractReranker`; `travsr-core`/`travsr-retrieval` remain ML-free (G3).
- [ ] `build_seed_set` reranks top-K after RRF; `Seed.rerank_score` populated; `rerank_budget_ms` circuit-breaker.
- [ ] **G1:** exact-symbol query (`SeedSource::Exact` present) is NOT governed by the rerank floor — deterministic exact-match confidence preserved (regression test).
- [ ] `classify_confidence` uses the absolute-floor gate on NL queries; `Calibration`/contamination-gate removed from the confidence path; embed-on/off branching collapsed.
- [ ] **G2:** confidence identical with embeddings on vs off for the same query.
- [ ] `get_context "delete all user accounts and drop the database"` → abstain, not STRONG; `"how is the knapsack budget optimizer implemented"` → Strong with correct seeds.
- [ ] Per-node display score is the absolute rerank score (no forced 1.00) in `get_context`.
- [ ] `travsr ask` parity: reproduction query → `matched: false`; `Score` column = absolute rerank score for seed rows, capped/tagged for expanded rows; `AskPayload.confidence` populated + printed.
- [ ] `ask` and `get_context` return the same confidence label + abstain decision (parity test).
- [ ] **G5:** daemon survives a forced inference panic (fail-open to lexical gate); model-load failure does not break startup.
- [ ] **G7:** `bench/` negative set added; floors recorded with model version; precision up on off-topic/broad queries, no recall regression on symbol queries.
- [ ] Reranker-off path (`TRAVSR_NO_RERANK=1` / warming / timeout) is byte-for-byte today's behaviour on both surfaces.

## Dependencies / scope

- **No dependency on travsr-embed #8 or #6** for v1 (in-process). The sidecar rerank (#8) is a future GPU/cloud variant behind the same `Reranker` trait.
- Out of scope: reranking the deterministic structural tools (`get_callers`/`get_dependencies`/`get_blast_radius`); reranker code-domain fine-tune (RFC-021 follow-up); MCP-sampling LLM judge (separate RFC).

Ref: RFC-021 (`docs/rfcs/RFC-021-cross-encoder-relevance-arbiter.md`), §15 for the full review.

---

## Delivery status (2026-07-24, E2E QA pass)

**Code:** Phases 0–5 + review fixes F1–F9 + dogfooding bugs A–D + WS1 all committed on `feature/travsr-rerank-rfc021-phase1` (PR #460). **Model distribution live:** immutable release `rerank-model-v1` published; `travsr rerank install` fetches + sha256-verifies both assets (~8s), F8 `model.toml` self-heals, daemon status transitions `installed → ready`.

**E2E verified working:** flagship fix closed — `delete all user accounts and drop the database` → abstains (none). Exact→`exact` (G1 bypass), grounded symbol/conceptual→`strong` on correct anchors (`impl:Daemon` 0.9065). F9 display shows real rerank scores (no "always 1.00"); ask/get_context confidence + abstain **parity confirmed** on both abstain and grounded cases. `TRAVSR_NO_RERANK=1` falls back cleanly.

**Known live limitations (do not block merge; tracked separately):**
- Verb-collision false-positive → #462 (higher-severity than recorded: `load the user's shopping cart and checkout` scores `fn:load_config` at 1.0 → false **strong**, not the "weak" documented in the plan; floor-tuning cannot fix a 1.0 FP).
- Vague-NL false-negative (`give me data about daemon` → abstain) → #462 / seed-diversity → #463.
